### PR TITLE
feat: cairo 1 migration

### PR DESCRIPTION
**Pull request by Orland0x**
This PR implements:
- Space state setters and getters.
- Propose, vote, update, execute cancel functions.
- Vanilla versions of each module.
- Eth tx authenticator.
- Eth Sig authenticator.
- L1 execution via an avatar contract.
- A small number of unit tests using the space contract and the vanilla modules

### CODE REVIEW

File: src/authenticators/vanilla.cairo [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+}
+
+#[contract]
+mod VanillaAuthenticator {
+    use starknet::ContractAddress;
+    use starknet::syscalls::call_contract_syscall;
+    use core::array::{ArrayTrait, SpanTrait};
+
+    #[external]
+    fn authenticate(target: ContractAddress, entry_point_selector: felt252, data: Array<felt252>) {
```
Comment (arr00) [inline src/authenticators/vanilla.cairo]: Why not refer to this as `selector` as in ethereum?

File: src/external/ownable.cairo [hunk 2]
```diff
@@ -0,0 +1,54 @@
+// OZ ownable module, adding here directly until OZ releases a Scarb package
```
Comment (arr00) [inline src/external/ownable.cairo]: No external function to transfer ownership?

File: src/execution_strategies/vanilla.cairo [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+}
+
+#[contract]
+mod VanillaExecutionStrategy {
+    use starknet::ContractAddress;
+    use sx::utils::types::Proposal;
+
+    struct Storage {
+        _num_executed: felt252
+    }
```
Comment (arr00) [inline src/execution_strategies/vanilla.cairo]: Is there an autogenerated getter? If not would be useful to confirm executions.

File: src/utils/bits.cairo [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+impl U256BitSetter of BitSetter<u256> {
+    /// Sets the bit at the given index to 1.
+    #[inline(always)]
+    fn set_bit(self: u256, index: u8, bit: bool) -> () {
+        let mask: u128 = pow(2, index.into());
+        let mask: felt252 = mask.into();
+        let mask: u256 = mask.into();
+        if bit {
+            self | mask;
```
Comment (arr00) [inline src/utils/bits.cairo]: I assume we want to return the expression `self | mask`. I believe the semicolon stops this from happening

File: src/utils/bits.cairo [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        // self & mask;
+        }
+    }
+
+    /// Returns true if the bit at the given index is set to 1.
+    #[inline(always)]
+    fn is_bit_set(self: u256, index: u8) -> bool {
+        let mask: u128 = pow(2, index.into());
+        let mask: felt252 = mask.into();
+        let mask: u256 = mask.into();
```
Comment (arr00) [inline src/utils/bits.cairo]: Confused how putting the mask into a `u128` then converting to `u256` would work

**Review by arr00 — COMMENTED** [review]
Just some intial comments

File: src/external/ownable.cairo (hunk 2 above)
Comment (Orland0x) [reply src/external/ownable.cairo]: i think modules cannot have external functions in cairo1.
 I added external functions `transfer_ownership` and `renounce_ownership`

File: src/utils/bits.cairo (hunk 4 above)
Comment (Orland0x) [reply src/utils/bits.cairo]: thanks, actually realised there was a lot wrong there haha. i changed the function to accept a mutable reference which gets updated, so no return value.

File: src/utils/bits.cairo (hunk 5 above)
Comment (Orland0x) [reply src/utils/bits.cairo]: yeah ignore that, was some old test code i forgot to remove

**Review by pscott — APPROVED** [review]
tACK!

Finally! :D This will of course be an ongoing effort, since Cairo is still evolving and we have quite a few TODOs left :)

**Pull request merged**